xrmocap/core/evaluation/fourd_association_evaluation.py

```diff
@@ -0,0 +1,440 @@
+                                                                           idx]
+
+            # save ground truth kps3d
+            if gt_kps3d is None:
+                gt_kps3d = kps3d.numpy()[np.newaxis]
+            else:
+                gt_kps3d = np.concatenate(
+                    (gt_kps3d, kps3d.numpy()[np.newaxis]), axis=0)
+
+        pred_keypoints3d = Keypoints(
+            dtype='numpy',
+            kps=pred_kps3d,
+            mask=pred_kps3d[..., -1] > 0,
+            convention=self.pred_kps3d_convention,
+            logger=self.logger)
+        gt_keypoints3d = Keypoints(
+            dtype='numpy',
+            kps=gt_kps3d,
+            mask=gt_kps3d[..., -1] > 0,
+            convention=self.dataset.gt_kps3d_convention,
+            logger=self.logger)
+
+        mscene_keypoints_paths = []
+        scene_start_idx = 0
+        for scene_idx, scene_end_idx in enumerate(end_of_clip_idxs):
+            scene_keypoints = pred_keypoints3d.clone()
+            kps3d = scene_keypoints.get_keypoints()[
+                scene_start_idx:scene_end_idx + 1, ...]
+            mask = scene_keypoints.get_mask()[scene_start_idx:scene_end_idx +
+                                              1, ...]
+            scene_keypoints.set_keypoints(kps3d)
+            scene_keypoints.set_mask(mask)
+            npz_path = osp.join(self.output_dir,
+                                f'scene{scene_idx}_pred_keypoints3d.npz')
+            scene_keypoints.dump(npz_path)
+            mscene_keypoints_paths.append(npz_path)
+            scene_start_idx = scene_end_idx + 1
+
+        pred_keypoints3d_, gt_keypoints3d_, limbs = self.align_keypoints3d(
+            pred_keypoints3d, gt_keypoints3d)
+        self.calc_limbs_accuracy(pred_keypoints3d_, gt_keypoints3d_, limbs)
+        pck_50, pck_100, mpjpe, pa_mpjpe = self.evaluate(
+            pred_keypoints3d_, gt_keypoints3d_)
+
+        if self.dataset_visualization is not None:
+            self.dataset_visualization.pred_kps3d_paths = \
+                mscene_keypoints_paths
+            self.dataset_visualization.run(overwrite=overwrite)
+
+    def evaluate(self,
+                 pred_keypoints3d: Keypoints,
+                 gt_keypoints3d: Keypoints,
+                 pck_50_thres=50,
+                 pck_100_thres=100,
+                 scale=1000.) -> dict:
+        # There must be no np.nan in the pred_keypoints3d
+        mpjpe, pa_mpjpe, pck_50, pck_100 = [], [], [], []
+        n_frame = gt_keypoints3d.get_frame_number()
+        gt_kps3d = gt_keypoints3d.get_keypoints()[..., :3]
+        gt_kps3d_mask = gt_keypoints3d.get_mask()
+        pred_kps3d = pred_keypoints3d.get_keypoints()[..., :3]
+        pred_kps3d_mask = pred_keypoints3d.get_mask()
+        pred_kps3d_convention = pred_keypoints3d.get_convention()
+        gt_kps3d_convention = gt_keypoints3d.get_convention()
+        for frame_idx in range(n_frame):
```

Comment:
It seems that you are looping point by point, sum those points whose mask==1. Is there a better way for mask usage? We need a dicussion with @jiaqiAA

Reply:
For example:
```python
def compute_mpjpe(pred_keypoints: Keypoints,
                  gt_keypoints: Keypoints,
                  align=False,
                  reduction='mean'):
    ...
    mpjpe = dist(pred_kps3d, gt_kps3d) * (pred_mask and gt_mask)
    if reduction == 'mean':
        mpjpe = mpjpe.sum() / (pred_mask and gt_mask).sum()
    return mpjpe
```
If we use mask like this, there will be not so many for loops.

Reply:
Also, if some evaluation functions are used by more than one paper, we could create a folder `metric` at `xrmocap/core/evaluation`, and remove the old one in `utils/mvpose_utils.py`

Reply:
I will adopt it.